src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -401,6 +501,162 @@ private static void AssertNoImplicitInterpolatedStringHandlerConversions(Immutab
+                    || partCall.Arguments[0] is BoundLiteral);
+                builder.Add(
+                    partCall.Method.Name == BoundInterpolatedString.AppendLiteralMethod
+                    ? partCall.Arguments[0]
+                    : new BoundStringInsert(partCall.Syntax, partCall.Arguments[0], null, null, false)
+                );
+            }
+            return builder.ToImmutableAndFree();
+        }
+
+        private (ImmutableArray<BoundExpression> Calls, int IncreasedLiteralLength, int FilledHolesCount) OptimizeAppendFormattedCalls(ImmutableArray<BoundExpression> parts)
+        {
+            var result = ArrayBuilder<BoundExpression>.GetInstance();
+            var currentStringConst = PooledStringBuilder.GetInstance();
+            (BoundCall? AppendLiteralCall, BoundLiteral? StringLiteral) reusable = default;
+            BoundExpression? lastAppendReceiver = null;
+            int filledHolesCount = 0;
+            int increasedLiteralLength = 0;
+
+            BoundExpression createAppendLiteralCallFromBoundLiteral(BoundLiteral literal, BoundExpression receiver) =>
+                _factory.InstanceCall(receiver, BoundInterpolatedString.AppendLiteralMethod, literal);
+            BoundExpression createAppendLiteralCallFromString(string literal, BoundExpression receiver) =>
+                createAppendLiteralCallFromBoundLiteral(_factory.StringLiteral(literal), receiver);
+            void mergeReusableIntoBuilder()
+            {
+                if (reusable.StringLiteral.ConstantValueOpt != null)
+                {
+                    currentStringConst.Builder.Append(reusable.StringLiteral.ConstantValueOpt.StringValue);
+                }
+                reusable = default;
+            }
+            void mergeReusableIntoResult(BoundCall? call = null)
+            {
+                switch (reusable)
+                {
+                    case (not null, _):
+                        result.Add(reusable.AppendLiteralCall);
+                        break;
+                    case (_, not null):
+                        Debug.Assert(call?.ReceiverOpt is not null || lastAppendReceiver is not null);
+                        result.Add(createAppendLiteralCallFromBoundLiteral(reusable.StringLiteral, call?.ReceiverOpt ?? lastAppendReceiver!));
+                        break;
+                    default:
+                        if (currentStringConst.Length > 0)
+                        {
+                            Debug.Assert(call?.ReceiverOpt is not null || lastAppendReceiver is not null);
+                            result.Add(createAppendLiteralCallFromString(currentStringConst.ToStringAndFree(), call?.ReceiverOpt ?? lastAppendReceiver!));
+                            currentStringConst = PooledStringBuilder.GetInstance();
+                        }
+                        return;
+                }
+                reusable = default;
+            }
+
+            foreach (var part in parts)
```

Comment:
I did not get very far into reviewing this `foreach`. It's extremely complicated, with method-level state implicitly flowing around (I have absolutely no idea what `reusable` is or what it's tracking). It needs be deeply commented, and not method-level state tracking. I would recommend starting by clarifying what `reusable` actually means, and making the `merge` methods be static local functions, to force all state to be explicitly shown.

Reply:
@333fred

> no idea what reusable is

It is used to retain nodes of `AppendLiteral` method calls and string literals that have not been joined to others yet.
They can be reused (output as is) until merged to others.
Maybe I should not have to define it as a tuple form.

> making the merge methods be static local functions

The reason why I have not made it `static` is just because I have to use `_factory`. The other part is static.
Is it true that you would like me to convert it to an additional parameter of the local function?
I do not care about whether `_factory` is a captured field or a function parameter.

Reply:
Nodes that may be retained by `reusable` are *just previous* part of the loop variable `part`.